modules/drivers/lidar/lidar_robosense/src/driver/driver16.cpp

```diff
@@ -0,0 +1,194 @@
+  }
+
+  if (scan->firing_pkts_size() == 0) {
+    AINFO << "Get a empty scan from port: " << config_.firing_data_port();
+    return false;
+  }
+
+  scan->set_model( config_.model());
+  scan->set_mode( config_.mode());
+  scan->mutable_header()->set_frame_id( config_.frame_id());
+  scan->mutable_header()->set_lidar_timestamp(
+      apollo::cyber::Time().Now().ToNanosecond());
+
+  scan->set_basetime( basetime_);
+  AINFO << "time: " << apollo::cyber::Time().Now().ToNanosecond();
+  return true;
+}
+
+/** poll the device
+ *
+ *  @returns true unless end of file reached
+ */
+void Robosense16Driver::poll_positioning_packet(void) {
+  while (!apollo::cyber::IsShutdown() && running_.load()) {
+    NMEATimePtr nmea_time(new NMEATime);
+    if (!config_.use_gps_time()) {
+      basetime_ = 1;  // temp
+      time_t t = time(NULL);
+      struct tm ptm;
+      struct tm* current_time = localtime_r(&t, &ptm);
+      nmea_time->year = current_time->tm_year - 100;
+      nmea_time->mon = current_time->tm_mon + 1;
+      nmea_time->day = current_time->tm_mday;
+      nmea_time->hour = current_time->tm_hour;
+      nmea_time->min = current_time->tm_min;
+      nmea_time->sec = current_time->tm_sec;
+      AINFO << "frame_id:" << config_.frame_id() << "-F(local-time):"
+            << "year:" << nmea_time->year << "mon:" << nmea_time->mon
+            << "day:" << nmea_time->day << "hour:" << nmea_time->hour
+            << "min:" << nmea_time->min << "sec:" << nmea_time->sec;
+    } else {
+      while (true && running_.load()) {
+        // AINFO<<"11111->gps timeing...";
+        if ( positioning_input_ == nullptr) {
+          AERROR << " positioning_input_ uninited.";
+          return;
+        }
+
+        int rc = positioning_input_->get_positioning_data_packtet(nmea_time);
+        AINFO << "gprmc data rc: " << rc << "---rc is normal when it is 0";
+
+        tm pkt_time;
+        memset(&pkt_time, 0, sizeof(pkt_time));
+        pkt_time.tm_year = nmea_time->year + 100;
+        pkt_time.tm_mon = nmea_time->mon - 1;
+        pkt_time.tm_mday = nmea_time->day;
+        pkt_time.tm_hour = nmea_time->hour + 8;
+        pkt_time.tm_min = nmea_time->min;
+        pkt_time.tm_sec = 0;
+
+        uint64_t timestamp_sec = static_cast<uint64_t>(mktime(&pkt_time)) * 1e9;
+        uint64_t timestamp_nsec =
+            static_cast<uint64_t>(nmea_time->sec * 1e6 + nmea_time->msec * 1e3 +
+                                  nmea_time->usec) *
+                1e3 +
+            timestamp_sec;  // ns
+        AINFO << "first POS-GPS-timestamp: [" << timestamp_nsec << "]";
+        basetime_ = timestamp_nsec;
+        AINFO << "frame_id:" << config_.frame_id()
+              << "-T(gps-time):" << nmea_time->year << "-" << nmea_time->mon
+              << "-" << nmea_time->day << "  " << nmea_time->hour << "-"
+              << nmea_time->min << "-" << nmea_time->sec;
+        AINFO << "first POS-GPS-time: [" << nmea_time->year << "/"
```

Comment:
Why so many and maybe duplicate prints? Better write a convenient function to help readability.